File: solutions/javascript/perfect-numbers/1/perfect-numbers.js
//
// This is only a SKELETON file for the 'Perfect Numbers' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const classify = (n) => {
  if(n <= 0)
    throw new Error('Classification is only possible for natural numbers.');
  let s = 1, i = 2, r = n % i, q = ~~(n / i);
  while(i <= q){
    if(r === 0)
      s += q != i ? q + i : q;
    r = n % ++i;
    q = ~~(n / i);
  }
  return s === n && n !== 1? 'perfect' : s > n ? 'abundant' : 'deficient';
};

File: solutions/python/perfect-numbers/1/perfect_numbers.py
def classify(number):
    if number <= 0:
        raise ValueError("Classification is only possible for positive integers.")
    if number == 1:
        return "deficient"
    aliquot_sum = 1
    for i in range(2, int(number ** 0.5) + 1):
        if number % i == 0:
            aliquot_sum += i 
            next_factor = number / i
            if next_factor != i:
                aliquot_sum += next_factor
    if aliquot_sum == number:
        return "perfect"
    elif aliquot_sum < number:
        return "deficient"
    else:
        return "abundant"
